src/NUnitFramework/framework/Internal/TestNameGenerator.cs

```diff
@@ -0,0 +1,457 @@
+                    else if (d == double.MaxValue)
+                        display = "double.MaxValue";
+                    else if (d == double.MinValue)
+                        display = "double.MinValue";
+                    else
+                    {
+                        if (display.IndexOf('.') == -1)
+                            display += ".0";
+                        display += "d";
+                    }
+                }
+                else if (arg is float)
+                {
+                    float f = (float)arg;
+
+                    if (float.IsNaN(f))
+                        display = "float.NaN";
+                    else if (float.IsPositiveInfinity(f))
+                        display = "float.PositiveInfinity";
+                    else if (float.IsNegativeInfinity(f))
+                        display = "float.NegativeInfinity";
+                    else if (f == float.MaxValue)
+                        display = "float.MaxValue";
+                    else if (f == float.MinValue)
+                        display = "float.MinValue";
+                    else
+                    {
+                        if (display.IndexOf('.') == -1)
+                            display += ".0";
+                        display += "f";
+                    }
+                }
+                else if (arg is decimal)
+                {
+                    decimal d = (decimal)arg;
+                    if (d == decimal.MinValue)
+                        display = "decimal.MinValue";
+                    else if (d == decimal.MaxValue)
+                        display = "decimal.MaxValue";
+                    else
+                        display += "m";
+                }
+                else if (arg is long)
+                {
+                    long l = (long)arg;
+                    if (l == long.MinValue)
+                        display = "long.MinValue";
+                    else if (l == long.MinValue)
+                        display = "long.MaxValue";
+                    else
+                        display += "L";
+                }
+                else if (arg is ulong)
+                {
+                    ulong ul = (ulong)arg;
+                    if (ul == ulong.MinValue)
+                        display = "ulong.MinValue";
+                    else if (ul == ulong.MinValue)
+                        display = "ulong.MaxValue";
+                    else
+                        display += "UL";
+                }
+                else if (arg is string)
+                {
+                    var str = (string)arg;
+                    bool tooLong = str.Length > stringMax;
+                    int limit = stringMax - THREE_DOTS.Length;
+
+                    StringBuilder sb = new StringBuilder();
+                    sb.Append("\"");
+                    foreach (char c in (string)arg)
+                    {
+                        sb.Append(EscapeControlChar(c));
```

Comment:
I think it does call ToString() but not all Types override it so you get the object implementation.

I'll look at the classes you mention. My guess is that only StringBuilder displays the contents right now. I'd prefer to limit this particular pull to truncation of strings where we already display them. If, for example, we wanted XmlNode to display the actual XML involved rather than &lt;System.Xml.XmlNode&gt; that seems better as a separate issue and PR, and a smaller one at that.